lesson4/README.md

```diff
@@ -1,9 +1,93 @@
+
+1. [Shared Preferences](https://developer.android.com/guide/topics/data/data-storage.html#pref):
+
+   Useful for writing small amount of data as key-value pairs.
+2. [Firebase](https://www.firebase.com/docs/android/quickstart.html):
+
+   A cool way to store data, run tests and analytics, authenticate, and monetize. The quickstart page linked here gives you a very good idea about how to use Firebase.
+3. [Internal Storage](https://developer.android.com/guide/topics/data/data-storage.html#filesInternal):
+
+   This memory chunk is private by default and is only available to the app. This data gets deleted when the app is uninstalled from the Android device. You can use methods like `openFileInput()`, `read()`, and `close()` on a [FileOutputStream](https://developer.android.com/reference/java/io/FileOutputStream.html)
+4. [External Storage](https://developer.android.com/guide/topics/data/data-storage.html#filesExternal):
+
+   The data stored here is world-readable but the data itself can't be relied upon as the user can unmount the storage anytime they want. Be sure to handle in the code the case where the data is unavailable.
+   ```java
+    /* Checks if external storage is available for read and write */
+    public boolean isExternalStorageWritable() {
+        String state = Environment.getExternalStorageState();
+        if (Environment.MEDIA_MOUNTED.equals(state)) {
+            return true;
+        }
+        return false;
+    }
+    
+    /* Checks if external storage is available to at least read */
+    public boolean isExternalStorageReadable() {
+        String state = Environment.getExternalStorageState();
+        if (Environment.MEDIA_MOUNTED.equals(state) ||
+            Environment.MEDIA_MOUNTED_READ_ONLY.equals(state)) {
+            return true;
+        }
+        return false;
+    }
+   ```
+5. [Cache](https://developer.android.com/reference/android/content/Context.html#getCacheDir()):
+
+   If you have data which qualifies this: "it would be nice to save it somewhere instead of making users type but it is no big deal if the data gets deleted", use cache. Remember that you can't store large amounts or data (>1MB) without causing problems. Also, be responsible about how you use the cache memory. Clean the space you occupied after you are done.
+6. [SQLite databases]():
+
+   Think of SQL as a giant table of rows and columns where you can name the columns. You can also ask the table for certain values within those columns. You also have the ability to build complex queries that ask for multiple things within multiple columns. You can create multiple SQL databases for your app and all of them will be available to all parts of your app but won't be available to anything outside the app. SQL databases are also stored locally.
```

Comment:
Yes, and this seems to be a way http://stackoverflow.com/questions/19217835/can-an-android-app-connect-directly-to-an-online-mysql-database
You actually need to figure out how to host it, etc. Too much work to set up when they can just use firebase (if they need an online database)